zk_prover/src/chips/range/range_check.rs

```diff
@@ -0,0 +1,173 @@
+use halo2_proofs::arithmetic::Field;
+use halo2_proofs::circuit::{AssignedCell, Layouter, Value};
+use halo2_proofs::halo2curves::bn256::Fr as Fp;
+use halo2_proofs::plonk::{Advice, Column, ConstraintSystem, Error, Fixed, Selector};
+use halo2_proofs::poly::Rotation;
+
+use std::fmt::Debug;
+
+use super::utils::decompose_fp_to_bytes;
+
+/// Configuration for the Range Check Chip
+///
+/// # Type Parameters
+///
+/// * `N_BYTES`: Number of bytes in which the value to be checked should lie
+///
+/// # Fields
+///
+/// * `z`: Advice column for the value to be checked and its running sum.
+/// * `range`: Fixed column for the lookup table. It contains the values from 0 to 2^8 - 1.
+/// * `toggle_lookup_check`: Selector to toggle the lookup check.
+///
+/// Patterned after [halo2_gadgets](https://github.com/privacy-scaling-explorations/halo2/blob/main/halo2_gadgets/src/utilities/decompose_running_sum.rs)
+#[derive(Debug, Clone)]
+pub struct RangeCheckConfig<const N_BYTES: usize> {
+    pub z: Column<Advice>,
+    pub range: Column<Fixed>,
+    pub toggle_lookup_check: Selector,
+}
+
+/// Helper chip that verfiies that the value witnessed in a given cell lies within a given range defined by N_BYTES.
+/// For example, Let's say we want to constraint 0x1f2f3f4f to be within the range N_BYTES=4.
+///
+/// `z(0) = 0x1f2f3f4f`
+/// `z(1) = (0x1f2f3f4f - 0x4f) / 2^8 = 0x1f2f3f`
+/// `z(2) = (0x1f2f3f - 0x3f) / 2^8 = 0x1f2f`
+/// `z(3) = (0x1f2f - 0x2f) / 2^8 = 0x1f`
+/// `z(4) = (0x1f - 0x1f) / 2^8 = 0x00`
+///
+///  |     | z          |
+///  | --- | ---------- |
+///  | 0   | 0x1f2f3f4f |
+///  | 1   | 0x1f2f3f   |
+///  | 2   | 0x1f2f     |
+///  | 3   | 0x1f       |
+///  | 4   | 0x00       |
+///
+/// The column z contains the witnessed value to be checked at offset 0
+/// At offset i, the column z contains the value z(i+1) = (z(i) - k(i)) / 2^8 (shift right by 8 bits) where k(i) is the i-th decomposition big-endian of `value`
+/// The contraints that are enforced are:
+/// - z(i) - 2^8⋅z(i+1) ∈ lookup_u8 (enabled by toggle_lookup_check activate at offset [0, N_BYTES - 1])
+/// - z(N_BYTES) == 0
+#[derive(Debug, Clone)]
+pub struct RangeCheckChip<const N_BYTES: usize> {
+    config: RangeCheckConfig<N_BYTES>,
+}
+
+impl<const N_BYTES: usize> RangeCheckChip<N_BYTES> {
+    pub fn construct(config: RangeCheckConfig<N_BYTES>) -> Self {
+        Self { config }
+    }
+
+    /// Configures the Range Chip
+    pub fn configure(
+        meta: &mut ConstraintSystem<Fp>,
+        z: Column<Advice>,
+        range: Column<Fixed>,
+        toggle_lookup_check: Selector,
+    ) -> RangeCheckConfig<N_BYTES> {
+        meta.annotate_lookup_any_column(range, || "LOOKUP_MAXBITS_RANGE");
+
+        meta.lookup_any("range u8 check for word", |meta| {
```

Comment:
In Ethereum, the term `Word` usually represents 256 bits, but the `z_0` value is less than 256 bits.
So, how about using simply `z_value` instead of `word` in this context?

Reply:
Thanks for the correction ! Fixed in 969ac61